mono/metadata/threads.c

```diff
@@ -4996,6 +4998,68 @@ mono_thread_is_foreign (MonoThread *thread)
 	return info->runtime_thread == FALSE;
 }
 
+#ifndef HOST_WIN32
+static void
+threads_native_thread_join_lock (gpointer key, gpointer value)
+{
+	pthread_t thread = (pthread_t)key;
+	if (thread != pthread_self ()) {
+		MONO_ENTER_GC_SAFE;
+		/* This shouldn't block */
+		mono_threads_join_lock ();
+		mono_native_thread_join (thread);
+		mono_threads_join_unlock ();
+		MONO_EXIT_GC_SAFE;
+	}
+}
+inline static void
+threads_native_thread_join_nolock (gpointer key, gpointer value)
+{
+	pthread_t thread = (pthread_t)key;
+	MONO_ENTER_GC_SAFE;
+	mono_native_thread_join (thread);
+	MONO_EXIT_GC_SAFE;
+}
+
+inline static void
+threads_add_joinable_thread_nolock (gpointer key)
+{
+	g_hash_table_insert (joinable_threads, key, key);
+}
+#else
+static void
+threads_native_thread_join_lock (gpointer key, gpointer value)
+{
+	MonoNativeThreadId thread_id = (MonoNativeThreadId)(guint64)key;
+	HANDLE thread_handle = (HANDLE)value;
+	if (thread_id != GetCurrentThreadId () && thread_handle != NULL && thread_handle != INVALID_HANDLE_VALUE) {
+		MONO_ENTER_GC_SAFE;
+		/* This shouldn't block */
+		mono_threads_join_lock ();
+		WaitForSingleObject (thread_handle, INFINITE);
+		mono_threads_join_unlock ();
+		MONO_EXIT_GC_SAFE;
+		CloseHandle (thread_handle);
+	}
+}
+
+inline static void
+threads_native_thread_join_nolock (gpointer key, gpointer value)
+{
+	HANDLE thread_handle = (HANDLE)value;
+	MONO_ENTER_GC_SAFE;
+	WaitForSingleObject (thread_handle, INFINITE);
```

Comment:
Why can't we use `mono_native_thread_join` on both platforms?

Reply:
see above